Add support for the EDHOC protocol

nvt opened the pull request:
This PR adds support for EDHOC, an implementation of Ephemeral Diffie-Hellman Over COSE (EDHOC) (IETF RFC9528).

It is based on work by Lidia Pocero (original implementation), Christos Koulamas, Peter A Jonsson, Rikard Höglund, Marco Tiloca, Joel Höglund, Niclas Finne, and myself.

It updates and replaces PR #1349.

rikard-sics commented:
I wanted to write a short comment in this PR about recent interoperability testing I did with this implementation of EDHOC. This was done during the IETF 122 Hackathon.

I successfully tested the EDHOC client from this codebase towards the following two EDHOC implementations:
- [Python implementation](https://github.com/chrysn/aiocoap) with demo server available at demo.coap.amsuess.com
- [Java implementation](https://github.com/rikard-sics/contiki-ng/tree/edhoc)

The configuration used for the testing was the following:
```
Credential type: CCS
ID cred type: KID
Cipher suite: 2
Method: 3 (Java & Python)
Method: 0 (Java)
```

In order to test successfully with the Python implementation I submitted a small fix to this codebase, which has been incorporated.

nvt commented:
Thanks for doing the interoperability testing at the IETF meeting, @rikard-sics! It is good to know that the tests were successful with method 0 and 3.

I can also mention that the next steps for this work-in-progress PR is to switch to the new Contiki-NG CBOR implementation and to do some general source code improvements before it is ready for review.

kkrentz commented:
On constrained devices, the elliptic curve cryptography (ECC) will block other protothreads for a long time. This may cause the watchdog to reboot or the radio to miss duty cycles. The ECC API suggested in #2898 allows other prothreads to make progress while executing ECC operations.

nvt committed:
Add functions for retrieving the current position and the remaining number of bytes of a CBOR reader.

nfi committed:
cbor: add reader API to skip next CBOR object

nvt committed:
EDHOC: imported an implementation.

nvt committed:
EDHOC: imported an implementation.

nvt committed:
COSE: imported an implementation.

nvt committed:
COSE: imported an implementation.

nvt committed:
EDHOC: imported client and server examples.

nvt committed:
EDHOC: imported client and server examples.

nvt committed:
tests: imported EDHOC tests.

nvt committed:
tests: imported EDHOC tests.

nvt committed:
log: add an EDHOC log level.

nvt committed:
log: add an EDHOC log level.

nvt committed:
Add security protocols tests.

nvt committed:
Add security protocols tests.

nvt committed:
doc: add EDHOC documentation.

nvt committed:
doc: add EDHOC documentation.

kkrentz commented on os/lib/cbor.c:
```diff
@@ -465,5 +465,17 @@ cbor_read_map(cbor_reader_state_t *state)
+const uint8_t *
+cbor_get_position(cbor_reader_state_t *state)
+{
+  return state ? state->cbor : NULL;
```
Allowing `state` to be NULL is inconsistent with the other functions.

kkrentz commented on os/lib/cbor.c:
```diff
@@ -477,5 +477,76 @@ cbor_get_remaining(cbor_reader_state_t *state)
+bool
+cbor_skip_next(cbor_reader_state_t *state)
+{
+  cbor_major_type_t major_type;
```
These declarations could be moved closer to where they are used.

kkrentz commented on os/net/security/edhoc/ecc-uecc.h:
```diff
@@ -0,0 +1,96 @@
+ * using the public point coordinates (gx, gy) and the private key. The public point is first uncompressed,
+ * and the ECC shared secret is computed using the uECC library. The result is stored in the output buffer `ikm`.
+ */
+bool uecc_generate_ikm(const uint8_t *gx_in, const uint8_t *gy_in,
```
It would be nicer to use descriptive variable names instead of gx_in, gy_in.

kkrentz commented on os/net/security/edhoc/ecc-uecc.c:
```diff
@@ -0,0 +1,139 @@
+  uECC_set_rng(&RNG);
+  uint8_t public_key[64];
+  watchdog_periodic();
+  err = uECC_make_key(public_key, key->priv, curve.curve);
```
Couldn't you reuse `lib/ecc.h` here, at least by busy-waiting for the protothread to exit?

kkrentz commented on os/net/security/cose/cose.c:
```diff
@@ -0,0 +1,451 @@
+#define LOG_MODULE "COSE"
+#define LOG_LEVEL LOG_LEVEL_EDHOC
+
+MEMB(encrypt0_storage, cose_encrypt0_t, 1);
```
Is it really necessary for COSE to be stateful. For my purposes, I, e.g., defined this interface:

```
/**
 * Implements AES-CCM-16-64-128 as specified in RFC 8152.
 *
 * @param aad     Data to authenticate.
 * @param aad_len Length of @p aad in bytes.
 * @param m       Data to encrypt.
 * @param m_len   Length of @p m in bytes.
 * @param key     Symmetric key to use.
 * @param nonce   Nonce to use.
 * @param forward @c true when securing and @c false when unsecuring.
 *
 * @return        0 on error and non-zero on success.
 */
int cose_encrypt0_aes_ccm_16_64_128(
    const uint8_t *aad, size_t aad_len,
    uint8_t *m, size_t m_len,
    const uint8_t key[COSE_ALGORITHM_AES_CCM_16_64_128_KEY_LEN],
    const uint8_t nonce[COSE_ALGORITHM_AES_CCM_16_64_128_IV_LEN],
    bool forward);
```

nvt committed:
cbor: remove inconsistent NULL checks in get_position/get_remaining

nvt committed:
cbor: tighten declaration scope in cbor_skip_next

nvt committed:
edhoc: use descriptive parameter names in uecc_generate_ikm

nvt committed:
edhoc: use descriptive parameter names in uecc_generate_ikm

nvt committed:
edhoc: route ECDH through the Contiki-NG ECC driver

Replace the duplicated per-backend wrappers (ecc-uecc, ecc-cc2538) and
the EDHOC_CONF_ECC build-time selection with a thin synchronous adapter
on top of lib/ecc.h. Long-running operations from the ECC driver are
busy-waited to completion so the EDHOC protocol code can stay
synchronous; switching to genuine PT_THREAD propagation through the
EDHOC core can be done as a follow-up.

nvt committed:
edhoc: route ECDH through the Contiki-NG ECC driver

Replace the duplicated per-backend wrappers (ecc-uecc, ecc-cc2538) and
the EDHOC_CONF_ECC build-time selection with a thin synchronous adapter
on top of lib/ecc.h. Long-running operations from the ECC driver are
busy-waited to completion so the EDHOC protocol code can stay
synchronous; switching to genuine PT_THREAD propagation through the
EDHOC core can be done as a follow-up.

nvt committed:
edhoc: fix format strings for int32_t ead_label

nvt committed:
edhoc: fix format strings for int32_t ead_label

nvt committed:
cose: replace stateful API with stateless helpers

nvt committed:
cose: replace stateful API with stateless helpers

nvt committed:
edhoc: recover peer ephemeral y-coordinate from x in ECDH wrapper

nvt committed:
edhoc: recover peer ephemeral y-coordinate from x in ECDH wrapper

nvt committed:
edhoc: route ECDSA sign/verify through the ECC driver

Route COSE_Sign1 signing and verification through the Contiki-NG ECC
driver protothreads (ecc_sign/ecc_verify) with watchdog_periodic() in
the busy-wait loops, matching the existing ECDH pattern. This prevents
watchdog resets during software P-256 ECDSA operations on slow MCUs.

nvt committed:
edhoc: route ECDSA sign/verify through the ECC driver

Route COSE_Sign1 signing and verification through the Contiki-NG ECC
driver protothreads (ecc_sign/ecc_verify) with watchdog_periodic() in
the busy-wait loops, matching the existing ECDH pattern. This prevents
watchdog resets during software P-256 ECDSA operations on slow MCUs.

nvt committed:
edhoc: replace process spin-loops with direct function calls

The edhoc_server and edhoc_client_protocol process threads never
yielded, so the process_start + process_run spin-loops were just an
indirect way of calling a function. Replace them with direct calls.

nvt committed:
edhoc: replace process spin-loops with direct function calls

The edhoc_server and edhoc_client_protocol process threads never
yielded, so the process_start + process_run spin-loops were just an
indirect way of calling a function. Replace them with direct calls.

nvt committed:
edhoc: unify CoAP response validation in EDHOC client

nvt committed:
edhoc: unify CoAP response validation in EDHOC client

nvt committed:
edhoc: remove dead code across EDHOC modules

nvt committed:
edhoc: remove dead code across EDHOC modules

nvt committed:
edhoc: simplify key storage list iteration

nvt committed:
edhoc: simplify key storage list iteration

nvt committed:
edhoc: inline gen_gxy() into edhoc_generate_prk_2e()

nvt committed:
edhoc: inline gen_gxy() into edhoc_generate_prk_2e()

nvt committed:
edhoc: use distinct enums for protocol states and event types

nvt committed:
edhoc: use distinct enums for protocol states and event types

nvt committed:
edhoc: move internal includes out of edhoc-client.h

nvt committed:
edhoc: move internal includes out of edhoc-client.h

nvt committed:
edhoc: return actual error status from msg_1 handler

edhoc_handler_msg_1() returned 'err' instead of the error_status returned
by edhoc_check_err_rx_msg() on the error path. At that point 'err' is
always 0, since it is initialized to 0 and the only preceding assignment
(set_rx_msg) is returned early on when negative.

The handler therefore detected an incoming EDHOC error message and then
reported success, so the caller continued the handshake as if a valid
msg_1 had been received. Return the actual status.

The corresponding paths in edhoc_handler_msg_2() and edhoc_handler_msg_3()
already return error_status.

nvt committed:
edhoc: stop truncating received message size to 8 bits

The received message size was cast to uint8_t before being stored in
buffers.rx_sz, which is a uint16_t. This truncated the size of any
EDHOC message or block larger than 255 bytes. Drop the cast.

nvt committed:
edhoc: demote verbose protocol traces from INFO to DBG
